[Detekt Baseline Warnings] Plugins Woo Commerce Module - Resolve/Suppress Complexity & Naming Warnings

ParaskP7 opened the pull request:
Parent #2479
Partially Closes: #2486

This PR resolves/suppresses all `complexity` and `naming` related warnings for the `plugins:woocommerce` module:

- `37` x complexity
    1. `1` x ComplexCondition (Suppress: d4af78b4971a41870dfb488d84957032d08d150d)
    2. `10` x ComplexMethod (Resolve: 28c1c3d45d15d21a882d1df413ddde91ae5b093b + fbe20a26eb7eba54e911a1ec38279f4e530bd976 + fb080616cf0cfb80c1b6e02d84afc03c018f75b1 + Suppress: + ba6729396f3d27a58021b72e92d40cf8474c10f3)
    3. `4` x LargeClass (Suppress: 874a49e3a54b243c91e41c952d56226c56923297)
    4. `10` x LongMethod (Resolve: 721debe369111a66d92a802e2b65f71db53bc89e + 38fdf3b0094c9ec2de2363a7f66d212073f3dc69 + Suppress: 732726889efb9d87311e5ba54503f9676f4c362a)
    5. `6` x LongParameterList (Suppress: 86eb8cdebabfa974fe9e110f7fe88949e334a743)
    6. `6` x NestedBlockDepth (Resolve: ad2e0efead32383109862207c11634f6325aacad + Suppress: cfb3497395b0d436daebc96e4065d96aff549b3e)
- `99` x naming
    1. `9` x ConstructorParameterNaming (Suppress: b02d7ed90baff83bb7a270da786296297a471897)
    2. `1` x MemberNameEqualsClassName (Suppress: 5196de5f84ca59bb4ef95bbbb03687a685506426)
    3. `89` x VariableNaming (Resolve: 6df77dbdff4b3825f444ba3bbf245306253b99c4 + Suppress: 3865d165987092986d60b98e974d9ca755b09af9)

PS.1: I am randomly adding an engineer from the WCAndroid team as the main reviewers, that is, in addition to the @wordpress-mobile/owl-team team itself, since I just want someone from that team to sign-off on that change as well (or additionally to the 🦉 team).

PS.2: I recommend reviewing this PR commit-by-commit.

-----

To test:

- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough (especially the `detekt` check).
- However, if you really want to be thorough, you could smoke test the `example` and/or the `WCAndroid` apps to verify that everything works as expected.

PS: Please ignore the failing `connected-tests` check as they are unrelated to this PR. This check has been failing consistency for some time now due to tests being flaky or needing to be updated/fixed.

Merge instructions:

- [x] Wait for #2494 to be approved and merged to `trunk`.
- [x] Make sure this PR's base is automatically updated to `trunk`.
- [x] Update PR with latest `trunk`.
- [x] Remove `[Status] Not Ready for Merge` label.
- [x] Merge PR to `trunk`.

ParaskP7 committed:
Analysis: Suppress complex condition warn for wc product store

Description: "Complex conditions make it hard to understand which cases
lead to the condition being true or false. To improve readability and
understanding of complex conditions consider extracting them into
well-named functions or variables and call those instead."

For more info see: https://detekt.dev/docs/rules/complexity
#complexcondition

ParaskP7 committed:
Analysis: Resolve complex method warning for from database addos mapper

Description: "Complex methods are hard to understand and read. It might
not be obvious what side-effects a complex method has. Prefer splitting
up complex methods into smaller methods that are in turn easier to
understand. Smaller methods can also be named much clearer which leads
to improved readability of the code.

This rule uses McCabe's Cyclomatic Complexity (MCC) metric to measure
the number of linearly independent paths through a function's source
code (https://www.ndepend.com/docs/code-metrics#CC). The higher the
number of independent paths, the more complex a method is. Complex
methods use too many of the following statements. Each one of them adds
one to the complexity count."

For more info see: https://detekt.dev/docs/rules/complexity
#complexmethod

ParaskP7 committed:
Refactor: Extract length range method for from database addons mapper

ParaskP7 committed:
Analysis: Suppress complex method warning for woo plugin module

Description: "Complex methods are hard to understand and read. It might
not be obvious what side-effects a complex method has. Prefer splitting
up complex methods into smaller methods that are in turn easier to
understand. Smaller methods can also be named much clearer which leads
to improved readability of the code.

This rule uses McCabe's Cyclomatic Complexity (MCC) metric to measure
the number of linearly independent paths through a function's source
code (https://www.ndepend.com/docs/code-metrics#CC). The higher the
number of independent paths, the more complex a method is. Complex
methods use too many of the following statements. Each one of them adds
one to the complexity count."

For more info see: https://detekt.dev/docs/rules/complexity
#complexmethod

ParaskP7 committed:
Analysis: Resolve long method warning for remote addon mapper

Description: "Methods should have one responsibility. Long methods can
indicate that a method handles too many cases at once. Prefer smaller
methods with clear names that describe their functionality clearly.

Extract parts of the functionality of long methods into separate,
smaller methods."

For more info see: https://detekt.dev/docs/rules/complexity#longmethod

ParaskP7 committed:
Refactor: Use prepare range method for remote addons mapper input multi.

Just like the rest such functions are already using, this
'inputMultiplier(...)' funciton should also reuse the
'prepareRange(...)' function for its 'quantityRange' field instead of
defining the very same logic itself again.

ParaskP7 committed:
Refactor: Transform length range function to prepare range

This is done in order for this class to match and become consistent with
its related 'RemoteAddonMapper' class, at least in terms of function
naming and parameters.

ParaskP7 committed:
Analysis: Suppress long method warning for woo plugin module

Description: "Methods should have one responsibility. Long methods can
indicate that a method handles too many cases at once. Prefer smaller
methods with clear names that describe their functionality clearly.

Extract parts of the functionality of long methods into separate,
smaller methods."

For more info see: https://detekt.dev/docs/rules/complexity#longmethod

ParaskP7 committed:
Analysis: Suppress large class warnings for woo plugin module

Description: "This rule reports large classes. Classes should generally
have one responsibility. Large classes can indicate that the class does
instead handle multiple responsibilities. Instead of doing many things
at once prefer to split up large classes into smaller classes. These
smaller classes are then easier to understand and handle less things."

For more info see: https://detekt.dev/docs/rules/complexity#largeclass

ParaskP7 committed:
Analysis: Suppress long parameter list warnings for woo plugin module

Description: "Reports functions and constructors which have more
parameters than a certain threshold."

For more info see: https://detekt.dev/docs/rules/complexity
#longparameterlist

ParaskP7 committed:
Analysis: Resolve nested block depth warnings for woo plugin module

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Analysis: Suppress nested block depth warnings for woo plugin module

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Analysis: Suppress constructor parameter naming warnings for woo plugin

Description: "Reports constructor parameter names that do not follow the
specified naming convention."

For more info see: https://detekt.dev/docs/rules/naming
#constructorparameternaming

ParaskP7 committed:
Analysis: Suppress member name equals class name warning for wc model

Description: "This rule reports a member that has the same as the
containing class or object. This might result in confusion. The member
should either be renamed or changed to a constructor. Factory functions
that create an instance of the class are exempt from this rule."

For more info see: https://detekt.dev/docs/rules/naming
#membernameequalsclassname

ParaskP7 committed:
Analysis: Resolve variable naming warning for order stats rest client

Description: "Reports variable names that do not follow the specified
naming convention."

For more info see: https://detekt.dev/docs/rules/naming#variablenaming

ParaskP7 committed:
Analysis: Suppress variable naming warning for woo plugin module

Description: "Reports variable names that do not follow the specified
naming convention."

For more info see: https://detekt.dev/docs/rules/naming#variablenaming

kidinov commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductModel.kt:
```diff
@@ -263,6 +263,7 @@ data class WCProductModel(@PrimaryKey @Column private var id: Int = 0) : Identif
     /**
      * Parses the images json array into a list of product images
      */
+    @Suppress("NestedBlockDepth")
```
Not sure what was criteria for fixing or suppressing are, but in this case extracting this to another method would probably fix that complaint. The same for the rest of the suppressed methods in this class 

```
                Gson().fromJson(images, JsonElement::class.java).asJsonArray.forEach { jsonElement ->
                    with(jsonElement.asJsonObject) {
                        WCProductImageModel(this.getLong("id")).also {
                            it.name = this.getString("name") ?: ""
                            it.src = this.getString("src") ?: ""
                            it.alt = this.getString("alt") ?: ""
                            imageList.add(it)
                        }
                    }
                }
```

ParaskP7 replied:
👋 @kidinov and thank you for your comment!

FYI: I agree, I think I can get rid of this `NestedBlockDepth` suppression by actually trying to resolve this and the other 2 such occurrences of it in this class.

PS: It will not be as simple as you suggested as I am not in favor of passing empty list to methods that will then populate them and still use that list externally from the calling method (I try to avoid thinking about the [passed by value vs reference](https://www.educative.io/answers/pass-by-value-vs-pass-by-reference) problem when I can). 🤔 Thus, I didn't want to touch this one, but I found an alternative. I am on it... ⚙️

ParaskP7 replied:
👋 @kidinov !

I just resolved those 3 occurrences of `NestedBlockDepth`. ✅ Let me know what you think: 🙇 

1. `getImageList(...)` -> 444a24a30fb54d0d8be782a4074f230a3a0c9f01
2. `parseJson(...)` -> 028129e8b5d783f678f690fdd385a5fe5fb09929
3. `getTriplets(...)` -> 123e072b1d58aff9d5f01b95963888313120b814

ParaskP7 replied:
👋 @kidinov !

Did you have the chance to look at the above changes? 🙏

kidinov commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderDto.kt:
```diff
@@ -3,8 +3,9 @@ package org.wordpress.android.fluxc.network.rest.wpcom.wc.order
-@Suppress("PropertyName")
+@Suppress("PropertyName", "VariableNaming")
 class OrderDto : Response {
+    @Suppress("ConstructorParameterNaming")
```
I am a bit concerned about suppressing the whole class here. It means that the new code also might violate this rule

You could use regexp to replace this in one go probably:

```
val (.+)_(.+): String\? = null
```

->
```
@SerializedName(\"$1_$2"\) val $1\u$2: String\? = null
```

<img width="574" alt="image" src="https://user-images.githubusercontent.com/4923871/184605522-95efe7cc-b01b-4ae3-8db9-e3f375f9b0ef.png">

ParaskP7 replied:
👋 @kidinov, thanks for being considerate and for sharing your concern!

As per my [reply comment](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/b02d7ed90baff83bb7a270da786296297a471897#r81199473) here, I didn't do this refactor because I wasn't sure about it, if that is being done by design, and since this functionality relates to orders, I was being careful not to update it unnecessarily.

> I am a bit concerned about suppressing the whole class here. It means that the new code also might violate this rule

PS: Yes, you are right, I could have instead suppressed each field individually instead of the whole data class and thus this would make any such violating new code to be detectable. I can do that if you like. I just chose to go for the higher level suppression because I was seeing multiple field violating this rule and I didn't want to add as many suppress annotations, making this data class a bit less readable. I would have hoped that whoever add new code to this data class would see the suppression and understand what not to do, but this is just an assumption of mine, thus, I totally get your concern.

Having said the above and if you say that it is safe for us to resolve instead of suppress these warnings, I'll proceed doing so. Just let me know, and I'll do that straight away. 💯

kidinov replied:
The same answer, yes you are right, those might be public and used by the clients of the library, and renaming it is risky. Better to keep it then

ParaskP7 replied:
Deal, better keep it as is (at least for now), thanks for the extra confirmation on that @kidinov ! 🙇

kidinov reviewed (approved):
LGTM! I left a couple of comments, but they are optional and can be resolved as the code is touched by someone

ParaskP7 committed:
Refactor: Remove unused get image urls function from wc product model

ParaskP7 commented:
> LGTM! I left a couple of comments, but they are optional and can be resolved as the code is touched by someone

Thank you for all your comments @kidinov , I really appreciate it and I am working on resolving as many as I can (and as agreed)! 🥇 ⚙️ 🌟

ParaskP7 committed:
Refactor: Rename get image list by adding the or empty suffix to it

ParaskP7 committed:
Analysis: Resolve nested block depth warning for get image list

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Analysis: Resolve nested block depth warning for parse json

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Refactor: Rename get triplets by adding the or empty suffix to it

ParaskP7 committed:
Analysis: Resolve nested block depth warning for get triplets

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into analysis/woo-plugin-complexity-naming-detekt-baseline-warnings

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into analysis/woo-plugin-complexity-naming-detekt-baseline-warnings
